### DISCUSSION TITLE
Treat `Final` attributes differently in dataclass-likes

### DISCUSSION THREAD

**Discussion by Enegg** [microsoft/pyright#7337] [Ideas]
According to the docs on [type qualifiers](https://typing.readthedocs.io/en/latest/spec/qualifiers.html#semantics-and-examples) (specifically the `ImmutablePoint` example), attributes annotated with `Final` within class body are interpreted as:
- class-level constants, if they have a value assigned,
- instance-level constants, if the value is assigned in `__init__`.

Considering following definition:
```py
from dataclasses import dataclass, field
from typing import Final

@dataclass
class Example:
    foo: Final[int]
    bar: Final[int] = field()

obj = Example(1, 2)
obj.foo = 3
obj.bar = 4
```
In both cases an error is rightfully reported, but the resulting message reveals the assignment of `= field()` makes pyright treat the `bar` attribute as a `ClassVar`. Also the mention of `__set__` is weird.
```
Cannot assign member "foo" for type "Example"
  "foo" is declared as Final and cannot be reassigned
    Member "__set__" is unknown

Cannot assign member "bar" for type "Example"
  Member "bar" cannot be assigned through a class instance because it is a ClassVar
  "bar" is declared as Final and cannot be reassigned
    Member "__set__" is unknown
```
(This is on pylance `v2024.2.103 (pre-release)`)

Since assignment to fields is used to give them a default value, or to customize them via the `field()` function, **and** `Final` fields within dataclass-likes become normal instance attributes (unlike `ClassVar`s, which do not become fields), I believe it makes sense to treat `Final` attributes in dataclass-likes as instance-level constants, despite the assignment to them.

P.S. it seems that pyright/pylance does not treat the `dataclasses` module right within jupyter notebooks. I'll file an issue on that later.

**Comment by erictraut** [comment]
[PEP 557](https://peps.python.org/pep-0557/), which introduced dataclasses, isn't clear on the intended behavior here. Nor is the typing spec. I don't think it's a good idea to change pyright's behavior unless/until the intended behavior is clarified in the typing spec. If you're interested in spearheading a change in the typing spec, here is the process that the typing council (of which I'm a member) has established: https://github.com/python/typing-council#decisions. If/when there is consensus on the intended behavior and the typing spec has been changed accordingly, then we can introduce [conformance tests](https://htmlpreview.github.io/?https://github.com/python/typing/blob/main/conformance/results/results.html) for the behavior, and you all Python type checkers (including pyright) will likely adopt that behavior.
